File: node-version/src/probes/NetworkProbe.js
import net from 'net';
import dns from 'dns/promises';

import ProbeResult from '../model/ProbeResult.js';

class NetworkProbe {

    constructor(host, port, timeoutMs) {
        this.host = host;
        this.port = port;
        this.timeoutMs = timeoutMs;
    }

    async executeProbe() {

        const timestamp = new Date();

        let reachable = false;
        let success = false;
        let latencyMs = 0;

        try {

            await dns.lookup(this.host);

            reachable = true;

            const startTime = process.hrtime.bigint();

            await new Promise((resolve, reject) => {

                const socket = new net.Socket();

                const timeout = setTimeout(() => {
                    socket.destroy();
                    reject(new Error('Connection timeout'));
                }, this.timeoutMs);

                socket.connect(this.port, this.host, () => {

                    const endTime = process.hrtime.bigint();

                    latencyMs =
                        Number(endTime - startTime) / 1_000_000;

                    clearTimeout(timeout);

                    success = true;

                    socket.destroy();

                    resolve();
                });

                socket.on('error', (err) => {

                    clearTimeout(timeout);

                    socket.destroy();

                    reject(err);
                });
            });

        } catch (error) {

            success = false;

        }

        return new ProbeResult(
            timestamp,
            latencyMs,
            reachable,
            success
        );
    }
}

export default NetworkProbe;

File: src/main/java/org/example/network/NetworkProbe.java
package org.example.network;

import org.example.model.ProbeResult;

import java.net.InetAddress;
import java.net.InetSocketAddress;
import java.net.Socket;
import java.time.LocalDateTime;

public class NetworkProbe {

    private final String host;
    private final int port;
    private final int timeoutMs;

    public NetworkProbe(String host, int port, int timeoutMs) {
        this.host = host;
        this.port = port;
        this.timeoutMs = timeoutMs;
    }

    public ProbeResult executeProbe() {

        LocalDateTime timestamp = LocalDateTime.now();

        boolean reachable = false;
        boolean success = false;
        double latencyMs = 0.0;

        try {

            InetAddress inetAddress = InetAddress.getByName(host);
            reachable = inetAddress.isReachable(timeoutMs);

            long startTime = System.nanoTime();

            try (Socket socket = new Socket()) {

                socket.connect(
                        new InetSocketAddress(host, port),
                        timeoutMs
                );

                long endTime = System.nanoTime();

                latencyMs = (endTime - startTime) / 1_000_000.0;
                success = true;
            }

        } catch (Exception e) {

            success = false;

        }

        return new ProbeResult(
                timestamp,
                latencyMs,
                reachable,
                success
        );
    }
}